clang/include/clang/Tooling/DependencyScanning/ModuleDepCollector.h

```diff
@@ -136,9 +138,15 @@ struct ModuleDeps {
   /// determined that the differences are benign for this compilation.
   std::vector<ModuleID> ClangModuleDeps;
 
-  /// Compiler invocation that can be used to build this module. Does not
-  /// include argv[0].
-  std::vector<std::string> BuildArguments;
+  /// Get (or compute) the compiler invocation that can be used to build this
+  /// module. Does not include argv[0].
+  const std::vector<std::string> &getBuildArguments();
+
+private:
+  friend class ModuleDepCollectorPP;
+
+  std::variant<std::shared_ptr<CowCompilerInvocation>, std::vector<std::string>>
+      BuildInvocationOrArguments;
```

Comment:
The size of `ModuleDeps` used to be 200B. Storing `CowCompilerInvocation` directly would increase that to 360B, while just pointing to it keeps it at 208B. WDYT about the trade-offs?

Also, this should semantically be `std::unique_ptr`, since we're not sharing the ownership. But I'm not a fan of the move-only semantics that implies. I could wrap it in a unique-ptr-like type that's copyable?

Reply:
> Also, this should semantically be std::unique_ptr, since we're not sharing the ownership. But I'm not a fan of the move-only semantics that implies. I could wrap it in a unique-ptr-like type that's copyable?

Not sure I understand this - if you copy it you spend the same storage as if it was inlined, so it seems like either you want shared_ptr or inline storage.

I think shared_ptr is reasonable so this can be easily copied around.

Reply:
My concern is that if I use `std::shared_ptr<CowCompilerInvocation>`, `ModuleDeps` stops having value semantics. I guess that would be solved by making it a `std::shared_ptr<const CowCompilerInvocation>` to prevent mutation of the invocation.

Yes, both inlined and indirect storage use the same amount of storage (+- 8B), but I was thinking about this from memory locality perspective (storing the rarely used invocation away from frequently used data). But I guess this data structure is full of pointers to heap anyways, so that doesn't really matter.

I guess I'll just go with the inline approach to save on boilerplate.